zip-0222.rst

```diff
@@ -0,0 +1,203 @@
+key - a user creating a transaction that spends the output must prove that they have
+access to the spending key controlling the output. For Sprout shielded outputs this was
+part of the zero-knowledge proof; for transparent and Sapling shielded outputs, the user
+creates a signature with the spending key.
+
+In some situations, more complex predicates are required. Zcash does not yet have any form
+of shielded programmability, and thus users rely on the script system inherited from
+Bitcoin to encumber transparent outputs. But even this is insufficient as-is for some
+desired use cases. In particular, the Bolt protocol [#zip-draft-bolt]_ requires a custom
+opcode, because the script system does not have the necessary primitives; this requires a
+change to the consensus rules.
+
+Shielded programmability would require significant changes across the ecosystem. In the
+interim, we propose in this ZIP a transparent form of Zcash extensibility that could later
+be migrated to shielded programmability.
+
+
+Conventions
+===========
+
+We use the following notation as defined in the Zcash protocol specification
+[#spec-notation]_, reproduced here for convenience:
+
+- length(*S*) means the length of (number of elements in) *S*.
+
+- { *a* .. *b* } means the set or type of integers from *a* through *b* inclusive.
+
+- *a* || *b* means the concatenation of sequences *a* then *b*.
+
+We reproduce the following conversion function from ZIP 32 [#zip-0032]_:
+
+- I2LEOSP\ :sub:`l`\ (*k*) is the byte sequence *S* of length *l*/8 representing in little-endian order the
+  integer *k* in range {0..2\ :sup:`l`\ -1}.
+
+
+Specification
+=============
+
+Transparent extension
+---------------------
+
+A transparent extension is identified by a numeric ``type``. A transparent extension may
+also have several modes of operation, corresponding to different kinds of encumbrance
+within the extension's overall protocol.
+
+A transparent extension mode is defined by four properties (in addition to ``type``):
+
+- A numeric ``mode``.
+- A byte sequence ``predicate`` containing an encoding of the predicate that is
+  prefix-free within this ``(type, mode)``.
+- A non-malleable byte sequence ``witness`` containing an encoding of evidence
+  satisfying the predicate that is prefix-free within this ``(type, mode)``.
+- A deterministic verification algorithm ``tze_verify`` that takes as arguments ``mode``,
+  ``predicate``, ``witness``, and a context object providing deterministic public
+  information about the transaction as well as the block chain (up to and including the
+  block that the transaction is mined in). It returns ``true`` if the predicate is
+  satisfied, and ``false`` otherwise.
+
+ZIPs that define a new transparent extension must completely specify all four of these
+properties for each defined mode.
+
+The lengths of ``predicate`` and ``witness`` are not required to be constant length, but
+MUST be solely determined by the pair ``(type, mode)``.
```

Comment:
This might be too restrictive for some extensions. For example, if an extension has some k-of-n component, then the length of the encoding would depend on values stored within the encoding itself, in addition to `(type, mode)`. This would still be compatible with a prefix-free encoding, but nodes wouldn't be able to enforce length validity without parsing the encoding. Maybe this could be handled by extensions having a light-weight length-checking function that could be run at block parsing time, and then full contextual parsing happens later during predicate validation.

Reply:
Should this be deleted entirely, or is converting MUST to SHOULD more appropriate?

Reply:
I'm in favour of deleting it entirely.